File: backend/tools.py
"""
Tools for the LangChain agent.
Includes: current_time, web_search, web_scrape, and memory tools.
"""

from datetime import datetime
from langchain_core.tools import tool
from bs4 import BeautifulSoup
import httpx
import os

from memory import save_memory as mem_save, get_memory as mem_get, list_memories as mem_list


from datetime import datetime, timezone


@tool
def current_time() -> str:
    """
    Get the current date and time in both UTC (for ISO database queries) and local timezone.
    
    Returns:
        Current UTC date/time (ISO 8601) and local timezone date/time.
    """
    now_utc = datetime.now(timezone.utc)
    now_local = datetime.now()
    tz_name = os.getenv("TZ", "America/Chicago")
    return (
        f"Current UTC time (for UTC ISO parameters): {now_utc.strftime('%Y-%m-%dT%H:%M:%SZ')}\n"
        f"Current Local time ({tz_name}): {now_local.strftime('%Y-%m-%d %H:%M:%S')}"
    )


@tool
def search_web(query: str) -> str:
    """
    Search the web using Tavily API.
    
    Args:
        query: The search query string.
        
    Returns:
        Search results with titles, snippets, and URLs.
    """
    try:
        from tavily import TavilyClient
        
        api_key = os.getenv("TAVILY_API_KEY")
        if not api_key:
            return "Tavily API key not configured. Please set TAVILY_API_KEY environment variable."
        
        client = TavilyClient(api_key=api_key)
        
        # Search with Tavily
        response = client.search(query, search_depth="basic", max_results=5)
        
        if not response.get("results"):
            return "No search results found."
        
        formatted = []
        for i, result in enumerate(response["results"], 1):
            title = result.get("title", "No title")
            snippet = result.get("content", "No description")
            url = result.get("url", "No URL")
            formatted.append(f"{i}. **{title}**\n   {snippet}\n   URL: {url}")
        
        return "\n\n".join(formatted)
    except Exception as e:
        return f"Search failed: {str(e)}"


import ipaddress
import socket
from urllib.parse import urlparse


def validate_scrape_url(url_str: str) -> None:
    """Validates URL to prevent SSRF against loopback, link-local, or private IP ranges."""
    allow_internal = os.getenv("ALLOW_INTERNAL_SCRAPE", "").lower() in ("true", "1", "yes")
    if allow_internal:
        return

    parsed = urlparse(url_str)
    if parsed.scheme.lower() not in ("http", "https"):
        raise ValueError(f"Forbidden URL scheme '{parsed.scheme}' (only http and https allowed).")

    hostname = parsed.hostname
    if not hostname:
        raise ValueError("Invalid host in URL.")

    try:
        ip_list = socket.getaddrinfo(hostname, None)
    except Exception as e:
        raise ValueError(f"Could not resolve host {hostname}: {e}")

    for addr in ip_list:
        ip_str = addr[4][0]
        ip_obj = ipaddress.ip_address(ip_str)
        if ip_obj.is_loopback or ip_obj.is_link_local or ip_obj.is_private or ip_obj.is_multicast or ip_obj.is_unspecified:
            raise ValueError(f"Access to restricted or internal IP {ip_str} ({hostname}) is forbidden.")


@tool
def scrape_url(url: str) -> str:
    """
    Scrape content from a web page URL.
    
    Args:
        url: The URL to scrape.
        
    Returns:
        The text content extracted from the page.
    """
    try:
        validate_scrape_url(url)
    except Exception as e:
        return f"Scraping blocked by security policy: {str(e)}"

    try:
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
        }
        
        max_bytes = 2097152  # Default 2MB limit
        env_max = os.getenv("MAX_SCRAPE_BYTES", "").strip()
        if env_max.isdigit() and int(env_max) > 0:
            max_bytes = int(env_max)

        response_content = bytearray()
        with httpx.stream("GET", url, headers=headers, timeout=30.0, follow_redirects=True) as response:
            response.raise_for_status()
            for chunk in response.iter_bytes():
                response_content.extend(chunk)
                if len(response_content) >= max_bytes:
                    break

        soup = BeautifulSoup(response_content.decode("utf-8", errors="ignore"), "html.parser")
        
        # Remove script and style elements
        for script in soup(["script", "style", "noscript"]):
            script.decompose()
        
        # Extract text
        text = soup.get_text(separator="\n", strip=True)
        
        # Clean up whitespace
        lines = text.split("\n")
        cleaned_lines = [line.strip() for line in lines if line.strip()]
        cleaned_text = "\n".join(cleaned_lines)
        
        # Limit to reasonable length
        if len(cleaned_text) > 5000:
            cleaned_text = cleaned_text[:5000] + "\n\n... (content truncated)"
        
        return f"Content from {url}:\n\n{cleaned_text}"
    except httpx.RequestError as e:
        return f"Failed to fetch URL: {str(e)}"
    except Exception as e:
        return f"Scraping failed: {str(e)}"


@tool
def save_memory_tool(key: str, content: str) -> str:
    """
    Save information to memory storage.
    
    Args:
        key: A unique identifier for this memory.
        content: The content to store.
        
    Returns:
        Confirmation message with the memory key.
    """
    result = mem_save(key, content)
    return f"Memory saved as: {key}"


@tool
def get_memory_tool(key: str) -> str:
    """
    Retrieve stored information by key.
    
    Args:
        key: The memory key to look up.
        
    Returns:
        The stored content or a not found message.
    """
    return mem_get(key)


@tool
def list_memories_tool() -> str:
    """
    List all saved memory keys.
    
    Returns:
        A list of memory keys.
    """
    return mem_list()

File: backend_go/pkg/tools/tools.go
package tools

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"

	"jada-backend/pkg/memory"

	"github.com/PuerkitoBio/goquery"
)

type ToolDefinition struct {
	Name        string                 `json:"name"`
	Description string                 `json:"description"`
	Parameters  map[string]interface{} `json:"parameters"`
	Execute     func(args map[string]interface{}) (string, error)
}

func GetCurrentTime() (string, error) {
	nowUTC := time.Now().UTC()
	nowLocal := time.Now()
	tzName := os.Getenv("TZ")
	if tzName == "" {
		tzName = "America/Chicago"
	}

	return fmt.Sprintf(
		"Current UTC time (for UTC ISO parameters): %s\nCurrent Local time (%s): %s",
		nowUTC.Format("2006-01-02T15:04:05Z"),
		tzName,
		nowLocal.Format("2006-01-02 15:04:05"),
	), nil
}

type TavilySearchRequest struct {
	Query       string `json:"query"`
	SearchDepth string `json:"search_depth"`
	MaxResults  int    `json:"max_results"`
}

type TavilySearchResult struct {
	Title   string `json:"title"`
	Content string `json:"content"`
	URL     string `json:"url"`
}

type TavilyResponse struct {
	Results []TavilySearchResult `json:"results"`
}

func SearchWeb(query string) (string, error) {
	apiKey := os.Getenv("TAVILY_API_KEY")
	if apiKey == "" {
		return "Tavily API key not configured. Please set TAVILY_API_KEY environment variable.", nil
	}

	reqBody := TavilySearchRequest{
		Query:       query,
		SearchDepth: "basic",
		MaxResults:  5,
	}

	jsonBytes, err := json.Marshal(reqBody)
	if err != nil {
		return "", fmt.Errorf("failed to marshal tavily request: %w", err)
	}

	req, err := http.NewRequest("POST", "https://api.tavily.com/search", bytes.NewBuffer(jsonBytes))
	if err != nil {
		return "", fmt.Errorf("failed to create tavily request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", apiKey))

	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Sprintf("Search failed: %v", err), nil
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Sprintf("Search failed with status %d: %s", resp.StatusCode, string(body)), nil
	}

	var tavilyResp TavilyResponse
	if err := json.NewDecoder(resp.Body).Decode(&tavilyResp); err != nil {
		return "", fmt.Errorf("failed to decode tavily response: %w", err)
	}

	if len(tavilyResp.Results) == 0 {
		return "No search results found.", nil
	}

	var formatted []string
	for i, res := range tavilyResp.Results {
		title := res.Title
		if title == "" {
			title = "No title"
		}
		snippet := res.Content
		if snippet == "" {
			snippet = "No description"
		}
		url := res.URL
		if url == "" {
			url = "No URL"
		}

		formatted = append(formatted, fmt.Sprintf("%d. **%s**\n   %s\n   URL: %s", i+1, title, snippet, url))
	}

	return strings.Join(formatted, "\n\n"), nil
}

func ScrapeURL(urlStr string) (string, error) {
	req, err := http.NewRequest("GET", urlStr, nil)
	if err != nil {
		return fmt.Sprintf("Failed to create request: %v", err), nil
	}

	req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")

	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Sprintf("Failed to fetch URL: %v", err), nil
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Sprintf("Scraping failed with status: %d", resp.StatusCode), nil
	}

	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return fmt.Sprintf("Scraping failed: %v", err), nil
	}

	doc.Find("script, style, noscript").Each(func(i int, s *goquery.Selection) {
		s.Remove()
	})

	text := doc.Text()

	lines := strings.Split(text, "\n")
	var cleanedLines []string
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if trimmed != "" {
			cleanedLines = append(cleanedLines, trimmed)
		}
	}

	cleanedText := strings.Join(cleanedLines, "\n")

	if len(cleanedText) > 5000 {
		cleanedText = cleanedText[:5000] + "\n\n... (content truncated)"
	}

	return fmt.Sprintf("Content from %s:\n\n%s", urlStr, cleanedText), nil
}

func GetLocalTools() []ToolDefinition {
	return []ToolDefinition{
		{
			Name:        "current_time",
			Description: "Get the current date and time in both UTC (for ISO database queries) and local timezone.",
			Parameters: map[string]interface{}{
				"type":       "object",
				"properties": map[string]interface{}{},
			},
			Execute: func(args map[string]interface{}) (string, error) {
				return GetCurrentTime()
			},
		},
		{
			Name:        "search_web",
			Description: "Search the web using Tavily API.",
			Parameters: map[string]interface{}{
				"type": "object",
				"properties": map[string]interface{}{
					"query": map[string]interface{}{
						"type":        "string",
						"description": "The search query string.",
					},
				},
				"required": []string{"query"},
			},
			Execute: func(args map[string]interface{}) (string, error) {
				q, _ := args["query"].(string)
				return SearchWeb(q)
			},
		},
		{
			Name:        "scrape_url",
			Description: "Scrape content from a web page URL.",
			Parameters: map[string]interface{}{
				"type": "object",
				"properties": map[string]interface{}{
					"url": map[string]interface{}{
						"type":        "string",
						"description": "The URL to scrape.",
					},
				},
				"required": []string{"url"},
			},
			Execute: func(args map[string]interface{}) (string, error) {
				u, _ := args["url"].(string)
				return ScrapeURL(u)
			},
		},
		{
			Name:        "save_memory_tool",
			Description: "Save information to memory storage.",
			Parameters: map[string]interface{}{
				"type": "object",
				"properties": map[string]interface{}{
					"key": map[string]interface{}{
						"type":        "string",
						"description": "A unique identifier for this memory.",
					},
					"content": map[string]interface{}{
						"type":        "string",
						"description": "The content to store.",
					},
				},
				"required": []string{"key", "content"},
			},
			Execute: func(args map[string]interface{}) (string, error) {
				key, _ := args["key"].(string)
				content, _ := args["content"].(string)
				return memory.SaveMemory(key, content), nil
			},
		},
		{
			Name:        "get_memory_tool",
			Description: "Retrieve stored information by key.",
			Parameters: map[string]interface{}{
				"type": "object",
				"properties": map[string]interface{}{
					"key": map[string]interface{}{
						"type":        "string",
						"description": "The memory key to look up.",
					},
				},
				"required": []string{"key"},
			},
			Execute: func(args map[string]interface{}) (string, error) {
				key, _ := args["key"].(string)
				return memory.GetMemory(key), nil
			},
		},
		{
			Name:        "list_memories_tool",
			Description: "List all saved memory keys.",
			Parameters: map[string]interface{}{
				"type":       "object",
				"properties": map[string]interface{}{},
			},
			Execute: func(args map[string]interface{}) (string, error) {
				return memory.ListMemories(), nil
			},
		},
	}
}
